p2p/module.go

```diff
@@ -333,13 +350,62 @@ func (m *p2pModule) handlePocketEnvelope(pocketEnvelopeBz []byte) error {
 		return fmt.Errorf("decoding network message: %w", err)
 	}
 
+	if m.isNonceAlreadyObserved(poktEnvelope.Nonce) {
+		// skip passing redundant message to application layer
+		return nil
+	}
+
+	if err := m.observeNonce(poktEnvelope.Nonce); err != nil {
+		return fmt.Errorf("pocket envelope nonce: %w", err)
+	}
+
+	// NB: Explicitly constructing a new `PocketEnvelop` literal with content
+	// rather than forwarding `poktEnvelope` to avoid blindly passing additional
+	// fields as the protobuf type changes. Additionally, strips the `Nonce` field.
 	event := messaging.PocketEnvelope{
 		Content: poktEnvelope.Content,
 	}
 	m.GetBus().PublishEventToBus(&event)
 	return nil
 }
 
+// observeNonce adds the nonce to the deduper if it has not been observed.
+func (m *p2pModule) observeNonce(nonce utils.Nonce) error {
+	// Add the nonce to the deduper
+	if err := m.nonceDeduper.Push(nonce); err != nil {
+		return err
+	}
+	return nil
+}
+
+// isNonceAlreadyObserved returns whether the nonce has been observed within the
+// deuper's capacity of recent messages.
+// DISCUSS(#278): Add more tests to verify this is sufficient for deduping purposes.
+func (m *p2pModule) isNonceAlreadyObserved(nonce utils.Nonce) bool {
+	if contains := m.nonceDeduper.Contains(nonce); contains {
```

Comment:
NIT: If you do `!contains` and return `false`, you wouldn't have to indent all the business logic.

Reply:
Thanks for calling this out! :raised_hands: I hadn't taken the time to properly refactor this function. :sweat_smile: